Make USBHIDKeyboard work at boot

RefactorFactory opened the pull request:
Add a boot_protocol parameter to the USBHIDKeyboard constructor. When true:

1. The USB interface descriptor will have an interface subclass of boot and
   and interface protocol of keyboard. This will cause some PC BIOS to send a
   a SET PROTOCOL BOOT request to the device.

2. The USB report descriptor will not have a report ID because boot protocol
   does not use report IDs.

3. When the host sends reports to the device, USBHIDKeyboard will accept a
   report ID of 0 instead of checking for HID_REPORT_ID_KEYBOARD(1).

4. When the device sends reports to the host, it will not send a report ID
   because boot protocol does not use report IDs.

5. Use endpoint address of 1 for input and output because some PC BIOS require
   it.

*By completing this PR sufficiently, you help us to review this Pull Request quicker and also help improve the quality of Release Notes*

### Checklist
1. [x] Please provide specific title of the PR describing the change, including the component name (*eg. „Update of Documentation link on Readme.md“*)
2. [ ] Please provide related links (*eg. Issue which will be closed by this Pull Request*)
3. [ ] Please **update relevant Documentation** if applicable
4. [x] Please check [Contributing guide](https://docs.espressif.com/projects/arduino-esp32/en/latest/contributing.html)

*This entire section above can be deleted if all items are checked.*

-----------
## Description of Change

USBHIDKeyboard didn't work at boot on my devices, so I've prototyped changes that make it work. I'm looking for feedback.

This adds a parameter to the USBHIDKeyboard constructor and when it is true, behavior is changed. I considered removing the parameter and changing the behavior all the time, but it seemed too risky.

This also adds a new parameter to the USBHID constructor, but it defaults to the old behavior.

I considered only using the new behavior if the host sends the SET PROTOCOL BOOT request, but it turns out that some BIOS do not send that request.

It is strange that some PC BIOS require specific endpoint addresses of 1 for input and output. I tried to do the reservation of the endpoint addresses similar to how USBCDC does it.

Some of the code seems ugly, so I'm looking for feedback.


## Tests scenarios

I tested this with Arduino-esp32 core v2.0.4 with an ESP32 S2 board, with one old laptop and one new desktop, both from major PC manufacturers.

## Related links

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=6964) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=6964) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/arduino-esp32?pullRequest=6964) it.</sub>

RefactorFactory commented on cores/esp32/esp32-hal-tinyusb.c:
```diff
@@ -661,6 +665,13 @@ esp_err_t tinyusb_enable_interface(tinyusb_interface_t interface, uint16_t descr
     }
+    if(interface == USB_INTERFACE_HID && reserve_endpoints){
+        // Some simple PC BIOS requires specific endpoint addresses for keyboard at boot
+        if(!tinyusb_reserve_out_endpoint(1) ||!tinyusb_reserve_in_endpoint(1)){
```
Without this reservation of endpoints IN 1, OUT 1, USBCDC will reserve OUT 3, IN 4, IN 5 (line 665), then when `USBHID.cpp:tusb_hid_load_descriptor()` calls `tinyusb_get_free_in_endpoint()`/`tinyusb_get_free_out_endpoint()`, those functions will return IN 3, OUT 4 because *those functions prefer returning endpoints whose opposite is already used*.

Some alternatives to calling `tinyusb_reserve_out_endpoint()`/`tinyusb_reserve_in_endpoint()` here to reserve IN 1, OUT 1:

1. `USBHID.cpp:tusb_hid_load_descriptor()` could call `tinyusb_get_free_duplex_endpoint()` which would likely get endpoints IN 1, OUT 1 because `USBHID.cpp:tusb_hid_load_descriptor()` runs early enough.
2. Change `tinyusb_get_free_in_endpoint()`/`tinyusb_get_free_out_endpoint()` not to prefer returning endpoints whose opposite is already used. Or add a flag parameter.

RefactorFactory commented:
Because USBHIDKeyboard is simple, these changes could be greatly simplified by doing some or all of the following:

1. Remove all the new flags and new parameters.
2. In the USB interface descriptor, **always** use an interface subclass of boot protocol.
3. **Never** use a report ID in the USB report descriptor or when receiving or sending reports.
4. **Always** use endpoints 1 IN, 1 OUT for USBHIDKeyboard.

I'm looking for feedback on whether some of these simplifications should be done.

RefactorFactory commented:
Any feedback on what I can do to get these changes in? I'm open to suggestions. Thanks.

RefactorFactory commented:
This has a problem on one of my PCs at boot where the PC does not send "SET PROTOCOL BOOT" to the device and it expects the device to send reports prefixed with a report ID. I'll investigate further.

me-no-dev commented:
USB Keyboards actually provide two HID devices, one for boot and another for all other cases. The non-boot device can have extra functions, buttons, etc.
I know tinyusb can have two devices, so looking into implementing that instead, seems like the better option here :)

RefactorFactory commented:
Ok, I've reworked this with some improvements.

I investigated a real keyboard from a major PC manufacturer and found that it makes the main HID device a boot keyboard. The real keyboard that I investigated has another HID device, but it is for special keys. So, like a real keyboard, I made `USBHIDKeyboard` use a descriptor that makes it a boot device, but it doesn't speak boot protocol (namely, omitting report IDs from reports) unless the host requests it.

This major PC manufacturer of a 2021 PC (which happens to not request boot protocol) needed two other work-arounds:

1. Use the endpoints 1 IN and 1 OUT. Without this, the device would never become ready at boot. Maybe there is some other deeper problem going on, but this work-around worked.
2. For reports using the shift key, break out the shift key state changes into separate reports. Without this, shifted characters would not be recognized at all (not even as non-shifted). I found this hard to believe, but this work-around worked. Because this work-around may not be necessary for everyone, I added an API, `USBHIDKeyboard::setShiftKeyReports()`, that can be called to disable this behavior, but left the default behavior enabled so that users don't have to spend hours rediscovering this problem.

Let me know any feedback, thanks.

RefactorFactory commented:
Any feedback on what I can do to get these changes in? I'm open to suggestions. Thanks.

RefactorFactory commented:
Any feedback on what I can do to get these changes in? I'm open to suggestions. Thanks.

me-no-dev commented:
sorry for the delay @RefactorFactory . We will look into this very soon and have it in what will become version 3.0.0 (based on IDF 5.x)

VojtechBartoska commented:
Hello @RefactorFactory, please sign CLA. Thanks

P-R-O-C-H-Y reviewed (approved):
Looks good. Can you please sign the CLA as requested? 
Thanks @RefactorFactory for the PR!

lucasssvaz reviewed (approved):
Works as expected. Tested on my ASRock motherboard (that didn't work without this PR).

RefactorFactory commented:
> Hello @RefactorFactory, please sign CLA. Thanks

> Looks good. Can you please sign the CLA as requested? Thanks @RefactorFactory for the PR!

I signed the CLA. Sorry it took so long. Thanks.
